### ISSUE TITLE
Rule proposal: Don't pass default values to functions

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#675]
# Fail

```js
function run(always = true) {}
run(true); // Error
```
```js
function run({always} = {always: true}) {}
run({always: true}); // Error
```
```js
function run({always = true} = {}) {}
run({always: true}); // Error
```
```js
function run(options) {
	options = {
		always: true,
		...options
	}
}
run({always: true}); // Error
```
```js
function run(options) {
	options = Object.assign({
		always: true
	}, options);
}
run({always: true}); // Error
```

And any other common defaults pattern (unless already blocked by other rules like [`prefer-object-spread`](https://eslint.org/docs/rules/prefer-object-spread))

# Pass

```js
run(); // No parameter set
```

# Exclude

If `run` depends on arity/`arguments.length`

**Comment by papb** [comment]
Your fourth example is not a "defaults pattern"... It is probably a mistake, since it overwrites the `always` option entirely... What you probably meant is:

```js
function run(options) {
	options = {
		always: true,
		...options
	}
}
run({always: true}); // Error
```

**Comment by fregante** [comment]
Heh, indeed. I rarely use that pattern so I always get it wrong

**Comment by papb** [comment]
Actually your last example is wrong too haha, should be:

```js
function run(options) {
	options = Object.assign({
		always: true
	}, options);
}
run({always: true}); // Error
```

**Comment by fregante** [comment]
That's what happens when I code in a `<textarea>` :D

**Comment by papb** [comment]
@fregante You forgot to fix it in the [ESLint issue](https://github.com/eslint/eslint/issues/13168) 😅

**Comment by fregante** [comment]
I'll hire you to follow my comments.

**Comment by papb** [comment]
@fregante 🤣

**Comment by fisker** [comment]
How much will you pay him? I only need half.

**Comment by fregante** [comment]
I never mentioned pay. It's all about exposure.

**Comment by sindresorhus** [comment]
While I like the idea of this rule, I feel it's not that useful when constrained to a single file. @fregante Maybe you could propose it over at https://github.com/typescript-eslint/typescript-eslint ?

**Comment by futpib** [comment]
This looks like something to be implemented in typescript itself.

**Comment by fregante** [comment]
TypeScript already supports this, partially. For example, if the default is `true`, then you only accept `false | undefined`: passing `true` will fail.

https://github.com/sindresorhus/refined-github/blob/d073d1480fa2b88e16632377b6835f93903716b2/source/features/index.tsx#L31

However this might be hard to manage for values that are passed around, e.g. if you can't assign a `boolean` coming from somewhere else.

**Comment by fregante** [comment]
I think what they said in https://github.com/typescript-eslint/typescript-eslint/issues/2304 means that this is not actually feasible (and I should close both issues)

Any last thoughts before I do?

**Comment by sindresorhus** [comment]
> I think what they said in typescript-eslint/typescript-eslint#2304 means that this is not actually feasible (and I should close both issues)

What I got from that is that it's not feasible because the TypeScript parser is not good enough. So maybe one day it will be feasible.

**Comment by fregante** [comment]
The TS-ESLint rule was rejected as "out of scope" there. Want to reopen here?

**Comment by sindresorhus** [comment]
The arguments have not changed though. Not working across files have limited use, and working across files is very difficult, even with the TypeScript parser. What do you propose?

**Comment by fregante** [comment]
I don't know the details of the implementation and I don't know why this could not work, given that there are rules that work across files

**Comment by sindresorhus** [comment]
> What I got from that is that it's not feasible because the TypeScript parser is not good enough. So maybe one day it will be feasible.

⬆️ 

It's not about running across files, but rather that it requires type information, and the type information provided is not detailed enough for this purpose.

**Comment by sindresorhus** [comment]
We could handle simple cases where they are in the same file. I don't think that happens a lot though.

**Comment by fregante** [comment]
> the type information provided is not detailed enough for this purpose.

Oh indeed.

 eslint-plugin-import kind of does this across files but with disastrous performance and complexity. I get it now
